[ALLUXIO-2117] Add caching in user to groups mapping.

ifcharming opened the pull request:
This PR implements the 1) solution in https://alluxio.atlassian.net/browse/ALLUXIO-2117

Implement a cache for user to groups mapping, which is refreshed every time interval (by default 1 min). Once the system admin updates user to groups mapping, it needs 1 min to propagate to take effect in Alluxio namespace.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10930/<h2>Build result: ABORTED</span></h2>[...truncated 2834 lines...]	at org.jvnet.hudson.maven3.launcher.Maven31Launcher.main(Maven31Launcher.java:132)	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)	at java.lang.reflect.Method.invoke(Method.java:606)	at org.codehaus.plexus.classworlds.launcher.Launcher.launchStandard(Launcher.java:330)	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:238)	at jenkins.maven3.agent.Maven31Main.launch(Maven31Main.java:181)	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)	at java.lang.reflect.Method.invoke(Method.java:606)	at hudson.maven.Maven3Builder.call(Maven3Builder.java:136)	at hudson.maven.Maven3Builder.call(Maven3Builder.java:71)	at hudson.remoting.UserRequest.perform(UserRequest.java:120)	at hudson.remoting.UserRequest.perform(UserRequest.java:48)	at hudson.remoting.Request$2.run(Request.java:326)	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)	at java.util.concurrent.FutureTask.run(FutureTask.java:262)	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)	at java.lang.Thread.run(Thread.java:745)[INFO] [INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ alluxio-tests ---
Test FAILed.

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+ */
+public class GroupMapping {
+  /** The underlying implementation of GroupMappingService. */
+  private final GroupMappingService mEnv;
```
s/mEnv/mService

ifcharming replied:
done

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+   * Default constructor to initialize cache.
+   */
+  public GroupMapping() {
+    mEnv = GroupMappingService.Factory.getUserToGroupsMappingService();
```
the Factory method should be called just `get`

ifcharming replied:
done

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+    mEnv = GroupMappingService.Factory.getUserToGroupsMappingService();
+    long timeoutMs = Long.parseLong(Configuration.get(
+        Constants.SECURITY_GROUP_MAPPING_CACHE_TIMEOUT_MS));
+    mCacheEnabled = timeoutMs > 0;
```
the fact that timeout == 0 means caching is disabled is not documented anywhere

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+public class GroupMapping {
+  /** The underlying implementation of GroupMappingService. */
+  private final GroupMappingService mEnv;
+  /** Whether the cache is enabled. Set timeout to non-positive value to disable cache */
```
end sentence with `.`

ifcharming replied:
done

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+  /**
+   * Invalidates and refreshes the cache.
+   */
+  public void refreshCache() {
```
this method does not seem to be used anywhere

ifcharming replied:
Removed, I was planning to add some shell command to enforce the `refreshCache`. We can always add back.

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class to map user to groups. It maintain a cache for user to groups mapping. The underlying
```
maintains

ifcharming replied:
done

jsimsa commented on core/common/src/main/java/alluxio/Constants.java:
```diff
@@ -474,6 +474,8 @@
       "alluxio.security.authorization.permission.supergroup";
   // Group Mapping
   public static final String SECURITY_GROUP_MAPPING = "alluxio.security.group.mapping.class";
```
this should be calls `SECURITY_GROUP_MAPPING_CLASS`

ifcharming replied:
done

jsimsa commented:
@ifcharming thanks for the PR, left some comments ... I also have a high-level question ... have you considered adding this logic into the `GroupMappingService` class itself? ... that would require changing the class from an interface to an abstract class, but it makes more sense to me to tightly couple the caching aspect with what you refer to as "implementation" ... that way we avoid having GroupMapping and GroupMappingService

gpang commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+        Constants.SECURITY_GROUP_MAPPING_CACHE_TIMEOUT_MS));
+    mCacheEnabled = timeoutMs > 0;
+    if (mCacheEnabled) {
+      mCache = CacheBuilder.newBuilder()
```
should there be a maximum size?

ifcharming replied:
added, default of 10000, should be enough for a typical organization

gpang commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+    mCacheEnabled = timeoutMs > 0;
+    if (mCacheEnabled) {
+      mCache = CacheBuilder.newBuilder()
+          .refreshAfterWrite(timeoutMs, TimeUnit.MILLISECONDS)
```
Could you leave a brief comment on what `refreshAfterWrite` and what `expireAfterWrite` both mean?

ifcharming replied:
added.

gpang commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+  /**
+   * Class to reload cache from underlying user group mapping service implementation.
+   */
+  private class GroupMappingCacheLoader extends CacheLoader<String, List<String>> {
```
The docs suggest making `reload` asynchronous: http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/cache/CacheLoader.html#reload(K,%20V)

Would that be possible?

ifcharming replied:
Good point! I've addd a async `reload`.

gpang commented:
@ifcharming Thanks for this fix! I left a few comments.

apc999 commented on core/common/src/main/java/alluxio/security/group/GroupMapping.java:
```diff
@@ -0,0 +1,95 @@
+ * Class to map user to groups. It maintain a cache for user to groups mapping. The underlying
+ * implementation depends on {@link GroupMappingService}.
+ */
+public class GroupMapping {
```
NIT: how about calling this class `GroupMappingService`? Up to you

apc999 replied:
and make this extending `GroupMappingService`

ifcharming replied:
Yeah, I've renamed to `CachedGroupMapping`, and re-organized the structure.

`GroupMappingService` is still the interface that developers can implement.
Its factory can provide `getCachedGroupMapping`

ifcharming commented:
@gpang @jsimsa  Thanks for the review! I've addressed your comments.

- Added async `reload`, and annotated `ReloadCache` params.
- Reorganized the structure of `CacheGroupMapping` and `GroupMappingService` interface, per discussion with Jiri offline.
- Make the `GroupMappingService` the centralized place for the global cache.
- After profiling, I've verified this caching approach indeed addressed performance bottleneck caused by Shell exec.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10932/<h2>Failed Tests: <span class='status-failure'>3</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/10932/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>3</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/10932/org.alluxio$alluxio-tests/testReport/alluxio.shell.command/LsCommandTest/lsTest/'><strong>alluxio.shell.command.LsCommandTest.lsTest</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/10932/org.alluxio$alluxio-tests/testReport/alluxio.shell.command/LsCommandTest/lsWildcardTest/'><strong>alluxio.shell.command.LsCommandTest.lsWildcardTest</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/10932/org.alluxio$alluxio-tests/testReport/alluxio.shell.command/LsCommandTest/lsrTest/'><strong>alluxio.shell.command.LsCommandTest.lsrTest</strong></a></li></ul>
Test FAILed.

apc999 commented on core/common/src/main/java/alluxio/security/group/GroupMappingService.java:
```diff
@@ -37,27 +37,68 @@
-    public static GroupMappingService getUserToGroupsMappingService() {
-      GroupMappingService mGroupMappingService;
+    public static GroupMappingService get() {
       if (LOG.isDebugEnabled()) {
```
`isDebugEnabled` seems unnecessary?

ifcharming replied:
done

apc999 commented on core/common/src/main/java/alluxio/security/group/CachedGroupMapping.java:
```diff
@@ -0,0 +1,126 @@
+  /** Create an executor service that provide ListenableFuture instances. */
+  private ThreadFactory mThreadFactory = new ThreadFactoryBuilder()
+      .setNameFormat("UserGroupMappingCachePool-%d").setDaemon(true).build();
+  private ExecutorService mParentExecutor = Executors.newSingleThreadExecutor(mThreadFactory);
```
NIT: `final`?

ifcharming replied:
done

apc999 commented on core/common/src/main/java/alluxio/security/group/CachedGroupMapping.java:
```diff
@@ -0,0 +1,126 @@
+  private static final long MAXSIZE = 10000;
+
+  /** Create an executor service that provide ListenableFuture instances. */
+  private ThreadFactory mThreadFactory = new ThreadFactoryBuilder()
```
NIT: `final`?

ifcharming replied:
done

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMappingService.java:
```diff
@@ -37,27 +37,68 @@
 
+    // TODO(chaomin): maintain a map from SECURITY_GROUP_MAPPING_CLASS name to cachedGroupMapping.
+    // Currently the single global cached GroupMappingService assumes that there is no dynamic
+    // configuration change for {@link Constants#SECURITY_GROUP_MAPPING_CALSS}.
```
s/CALSS/CLASS

ifcharming replied:
done

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMappingService.java:
```diff
@@ -37,27 +37,68 @@
+    // TODO(chaomin): maintain a map from SECURITY_GROUP_MAPPING_CLASS name to cachedGroupMapping.
+    // Currently the single global cached GroupMappingService assumes that there is no dynamic
+    // configuration change for {@link Constants#SECURITY_GROUP_MAPPING_CALSS}.
+    private static GroupMappingService sGroupMappingService = null;
```
why are we caching these?

ifcharming replied:
otherwise a new instance of `GroupMappingService` and `CachedGroupMapping` will be created every time `get` is called.

jsimsa replied:
which is what we were doing before this PR, no?

ifcharming replied:
Before this PR, yes, tons of `GroupMappingService` instances were created....
It was not a big deal because there was no cache for GroupMappingService at all. Now we want to keep a global cache so we should cache those two instances.

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMappingService.java:
```diff
@@ -37,27 +37,68 @@
+    /**
+     * Resets the cache for GroupMappingService and CachedGroupMapping.
+     */
+    public static void resetCache() {
```
not used?

ifcharming replied:
used in `AbstractLocalAlluxioCluster` to clear minicluster state.

jsimsa replied:
I think in that case it should be put in the test code

ifcharming replied:
done, thanks.

jsimsa commented on core/common/src/main/java/alluxio/security/group/GroupMappingService.java:
```diff
@@ -37,27 +37,68 @@
+     *
+     * @return the cached user-to-groups mapping object
+     */
+    public static CachedGroupMapping getCachedGroupMapping() {
```
why don't we do this by default?

ifcharming replied:
Well, I feel it's weird to return a different type `CachedGroupMapping` when you call `GroupMappingService.Factory.get()`. I'd expect a `GroupMappingService` is returned by default.

jsimsa replied:
CachedGroupMapping service implements GroupMappingService, no?

ifcharming replied:
yup, updated. thanks!

jsimsa commented on core/common/src/test/java/alluxio/security/authorization/PermissionTest.java:
```diff
@@ -87,7 +87,8 @@ public void setOwnerFromThriftClientTest() throws Exception {
 
     Configuration.set(Constants.SECURITY_AUTHENTICATION_TYPE, AuthType.SIMPLE.getAuthName());
-    Configuration.set(Constants.SECURITY_GROUP_MAPPING, IdentityUserGroupsMapping.class.getName());
+    Configuration.set(Constants.SECURITY_GROUP_MAPPING_CLASS,
```
why are all these lines changed?

jsimsa commented on core/client/src/test/java/alluxio/client/file/options/CompleteUfsFileOptionsTest.java:
```diff
@@ -43,7 +43,8 @@ public void defaultsTest() throws IOException {
     Configuration.set(Constants.SECURITY_LOGIN_USERNAME, "foo");
     // Use IdentityOwnerGroupMapping to map owner "foo" to group "foo".
-    Configuration.set(Constants.SECURITY_GROUP_MAPPING, IdentityUserGroupsMapping.class.getName());
+    Configuration.set(Constants.SECURITY_GROUP_MAPPING_CLASS,
```
is this needed?

apc999 commented on core/common/src/main/java/alluxio/security/group/GroupMappingService.java:
```diff
@@ -37,27 +37,68 @@
-        mGroupMappingService = CommonUtils.createNewClassInstance(
-            Configuration.<GroupMappingService>getClass(Constants.SECURITY_GROUP_MAPPING), null,
-            null);
+        if (sGroupMappingService == null) {
```
why we need to check `sGroupMappingService == null` twice here?

jsimsa replied:
to avoid the race

ifcharming replied:
Because  double-checked locking is better. After it grabs the lock, it should check null again in case the value is already set by another thread.

apc999 replied:
I guess I am asking why we need to sync here rather than initializing in the static block. later I realize that it is probably due to an effort to avoid make static init block depends on `Constant` or `Configuration`. I guess this can be solved later with my configuration change

apc999 replied:
as discussed offline, we prevent anything in static init block to depends on reading configuration. Might be worth to put a note here to prevent later contributors change this following my thought.

jsimsa commented on minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java:
```diff
@@ -423,6 +424,7 @@ public void run() {
    */
   protected void reset() {
     ClientTestUtils.resetClient();
+    GroupMappingService.Factory.resetCache();
```
This should be `GroupMappingTestUtils.resetCache();`

ifcharming replied:
done.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10934/
Test PASSed.

ifcharming commented:
@jsimsa @apc999  The PR is updated. PTAL, thanks.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10936/
Test PASSed.

jsimsa commented:
LGTM

apc999 commented:
LGTM
